node/src/WorkQueue.ts

```diff
@@ -61,7 +63,47 @@ export class WorkQueue {
   }
 
   /**
-   * This is used to get the length of the Main Queue.
+   * Adds an item to the work queue only if an item with the same ID doesn't already exist.
+   *
+   * This method uses WATCH to add the item atomically. The db client passed must not be used by
+   * anything else while this method is running.
+   *
+   * Returns a boolean indicating if the item was added or not. The item is only not added if it
+   * already exists or if an error (other than a transaction error, which triggers a retry) occurs.
+   *
+   * @param {Redis} db The Redis Connection, this must not be used by anything else while this method is running.
+   * @param item The item that will be added, only if an item doesn't already exist with the same ID.
+   * @returns {boolean} returns false if already in queue or true if the item is successfully added.
+   */
+  async addNewItem(db: Redis, item: Item): Promise<boolean> {
+    for (;;) {
+      try {
+        await db.watch(this.mainQueueKey, this.processingKey);
+
+        const isItemInProcessingKey = await db.lpos(this.processingKey, item.id);
+        const isItemInMainQueueKey = await db.lpos(this.mainQueueKey, item.id);
+        if (isItemInProcessingKey !== null || isItemInMainQueueKey !== null) {
+          await db.unwatch();
+          return false;
+        }
+
+        const transaction = db.multi();
+        this.addItemToPipeline(transaction, item);
+        const results = await transaction.exec();
```

Comment:
@Oxygen588 what does the return value look like in the success and failure case? I'm not sure what it would be, so it would be nice to have a comment!

Reply:
Sorry, to clarify what I mean, I don't know what the value of `results` if the transaction succeeds or fails.